src/backend/parser/parse_coerce.c

```diff
@@ -1507,32 +1508,80 @@ select_common_type(ParseState *pstate, List *exprs, const char *context,
 		}
 	}
 
-	/*
-	 * If all the inputs were UNKNOWN type --- ie, unknown-type literals ---
-	 * then resolve as type TEXT.  This situation comes up with constructs
-	 * like SELECT (CASE WHEN foo THEN 'bar' ELSE 'baz' END); SELECT 'foo'
-	 * UNION SELECT 'bar'; It might seem desirable to leave the construct's
-	 * output type as UNKNOWN, but that really doesn't work, because we'd
-	 * probably end up needing a runtime coercion from UNKNOWN to something
-	 * else, and we usually won't have it.  We need to coerce the unknown
-	 * literals while they are still literals, so a decision has to be made
-	 * now.
-	 */
-	if (ptype == UNKNOWNOID && sql_dialect == SQL_DIALECT_TSQL)
+
+	if (sql_dialect == SQL_DIALECT_TSQL)
 	{
-		bool all_nullconst = true;
```

Comment:
Is it possible to put all those logic into single hook and move to babelfish code only ?

Reply:
I've optimized the codes and nest to the original loop. And it works together with the following codes outside of the loop. So it's not easy to make them in a single hook.